packages/massa-web3/src/experimental/smartContract.ts

```diff
@@ -0,0 +1,233 @@
+      datastore.set(
+        new Uint8Array(
+          new Args()
+            .addU64(BigInt(i + 1))
+            .addUint8Array(u8toByte(0))
+            .serialize()
+        ),
+        new Uint8Array(contract.args)
+      )
+    }
+    if (contract.coins > 0) {
+      datastore.set(
+        new Uint8Array(
+          new Args()
+            .addU64(BigInt(i + 1))
+            .addUint8Array(u8toByte(1))
+            .serialize()
+        ),
+        u64ToBytes(BigInt(contract.coins))
+      )
+    }
+  })
+
+  return datastore
+}
+
+interface DeployOptions {
+  smartContractCoins?: number
+  fee?: number
+  periodToLive?: number
+  coins?: number
+  maxGas?: number
+  waitForSpeculativeExecution?: boolean
+}
+
+interface CallOptions {}
+
+/**
+ * A class to interact with a smart contract.
+ */
+export class SmartContract {
+  protected constructor(
+    public client: BlockchainClient,
+    public contractAddress: string
+  ) {}
+
+  /**
+   * Initializes a new smart contract object from a contract address.
+   *
+   * @param client - The client to connect to the desired blockchain.
+   * @param contractAddress - The address of the smart contract.
+   *
+   * @returns A new smart contract instance.
+   */
+  static fromAddress(
+    client: BlockchainClient,
+    contractAddress: string
+  ): SmartContract {
+    return new SmartContract(client, contractAddress)
+  }
+
+  /**
+   * Deploys a smart contract on the blockchain.
+   *
+   * @param client - The client to connect to the desired blockchain.
+   * @param account - The account that will deploy the smart contract.
+   * @param byteCode - The byte code of the smart contract.
+   * @param Args - The arguments of the smart contract constructor.
+   * @param opts - Optional deployment details.
+   *
+   * @returns The deployed smart contract.
+   *
+   * @throws If the account has insufficient balance to deploy the smart contract.
+   */
+  static async deploy(
+    client: BlockchainClient,
+    account: Account,
+    byteCode: Uint8Array,
+    Args: Uint8Array,
+    opts: DeployOptions
+  ): Promise<SmartContract> {
+    const totalCost =
+      Number(StorageCost.smartContract(byteCode.length)) + opts.coins
+
+    if (
+      (await client.getBalance(account.address.toString(), true)) < totalCost
+    ) {
+      throw new Error('Insufficient balance')
+    }
+
+    const datastore = populateDatastore([
+      {
+        data: byteCode,
+        args: Args,
+        coins: BigInt(opts.smartContractCoins),
+      },
+    ])
+
+    // TODO: fix the code to execute the deployer smart contract instead
+    const operation = await ByteCode.execute(
+      client,
+      account.privateKey,
+      byteCode,
+      {
+        fee: opts?.fee,
+        periodToLive: opts?.periodToLive,
+        coins: opts?.coins,
+        maxGas: opts?.maxGas,
+        datastore,
+      }
+    )
+
+    const event = opts.waitForSpeculativeExecution
+      ? await operation.getSpeculativeEvents()
```

Comment:
I find the boolean misleading. When I read it I think waiting for speculative or not waiting at all. I would prefer something like an enum with no wait, wait speculative, final. If we just want the last two options we can rename it `waitFinalExecution` if false : speculative, if true final

Reply:
Is the noWait a real use case?

Reply:
Let's go with renaming then.